app/src/main/java/org/hitlabnz/sensor_fusion_demo/Compass3D.java

```diff
@@ -0,0 +1,310 @@
+        label = new Label("FPS:", new LabelStyle(font, Color.YELLOW));
+
+        // add all UI elements to the table, each into a new row
+        table.add(label).top().left().pad(16f);
+        //table.row(); // only if we want to add new UI element to the table in a new row
+
+        // write some logs
+        Gdx.app.log(TAG, "App created!");
+        Gdx.app.log(TAG, "" + Gdx.app.getType());
+        Gdx.app.log(TAG, "GDX version: " + Version.VERSION);
+    }
+
+    @Override
+    public void resize(int width, int height) {
+        // should not happen on Android using the current settings,
+        // manifest file - hardcoded screen portrait mode,
+        // but anyway, on resize we update the camera viewport
+        cam.viewportWidth = width;
+        cam.viewportHeight = height;
+        cam.update();
+
+        stage.getViewport().update(width, height, true);
+    }
+
+    @Override
+    public void render() {
+        // clear screen
+        ScreenUtils.clear(Color.BLACK, true);
+
+        // render the model - our 3D compass
+        modelBatch.begin(cam);
+        modelBatch.render(instance, lights);
+        modelBatch.end();
+
+        // render some text information
+        batch.begin();
+        label.setText("FPS: " + Gdx.graphics.getFramesPerSecond() + "\nGDX version: " + Version.VERSION + "\nCurrent orientation provider:\n" + currentOrientationProvider);
+        stage.act(Gdx.graphics.getDeltaTime());
+        stage.draw();
+        batch.end();
+
+        // Rotate virtual camera based on device orientation
+        // get rotation matrix from the current orientation provider
+        activity.getOrientationProvider().getRotationMatrix(rm);
+        RotMat = rm.getMatrix();
+
+        // TODO: Handle Display Rotation using WindowManager.DefaultDisplay.Rotation
+        // Currently the activity is hardcoded set to portrait mode.
+        // When the device is held with screen facing to the user,
+        // the sensors coordinate system is different than the world coordinate system and
+        // therefore needs to be remapped to show the 3D compass correctly.
+        SensorManager.remapCoordinateSystem(RotMat, SensorManager.AXIS_X, SensorManager.AXIS_Z, ModMat);
+        SensorManager.getOrientation(ModMat, angles);
+
+        float Azimuth = (angles[0] * MathUtils.radiansToDegrees + 360f) % 360f; // azimuth
+        float Pitch = angles[1] * MathUtils.radiansToDegrees; // altitude
+        float Roll = angles[2] * MathUtils.radiansToDegrees; // camera roll
+
+        //Gdx.app.log(TAG, MessageFormat.format("Azimuth: {0,number,#.###}; Pitch: {1,number,#.###}; Roll: {2,number,#.###}", Azimuth, Pitch, Roll));
+
+        // Transform the Forward vector (Vector3.Z) using a quaternion got from Azimuth/Pitch angles to a LookAt vector
+        Quaternion rotCam = new Quaternion().setEulerAngles(-Azimuth, Pitch, 0);
+        Vector3 newDir = new Vector3(Vector3.Z).mul(rotCam).nor();
+        cam.lookAt(newDir);
```

Comment:
This whole conversion here seems a bit excessive. First we get the rotation matrix, then we convert it to euler angles, and then into a Quaternion, before converting it again into some rotation vector. What is the reason for these conversion?

Reply:
> This whole conversion here seems a bit excessive. First we get the rotation matrix, then we convert it to euler angles, and then into a Quaternion, before converting it again into some rotation vector. What is the reason for these conversion?

You missed one important step:
First we get the rotation matrix from provider and we immediately convert it to a new rotation matrix but with a remapped coordinate system.
This is due to the fact that the hardware sensors coordinate system is different than the 3D world coordinate system, when you hold the device and the screen is facing to you:
In this case the Y and Z axes are swapped, that's why we need a new rotation matrix.

Next when you read the code carefully, you can see that first we process only 2 Euler angles out of the 3: The Azimuth (Yaw, rotation around the world's Y axis) and the Pitch (rotation around the world's X axis).
In most of the cases (3D games) the virtual camera really uses only the rotations around the above two WORLD axes. This gives us a new camera DIRECTION, where the camera actually looks at.

In our case we also apply an additional rotation but this time around the previously calculated camera's DIRECTION vector which involves the third Euler angle, the Roll.

That's why the excessive computation with two calculated quaternions. One we use to adjust the camera's lookAt vector while the other to adjust the camera's UP vector.

If you know how to calculate this in a simpler and easier way, please tell me 🙃